coding_style/README.adoc

```diff
@@ -0,0 +1,217 @@
+  If you set on a block, it will ignore all the asserts in the block ultimately making them pointless.
+* Do not use the `eq` (introduced in Jinja 2.10) or `equalto` Jinja operators.
+* Avoid the use of `when: foo_result is changed` whenever possible.
+  Use handlers, and, if necessary, handler chains to achieve this same result.
+* Use the various include/import statements in Ansible.
++
+[%collapsible]
+====
+Explanation:: Doing so can lead to simplified code and a reduction of repetition.
+This is the closest that Ansible comes to callable sub-routines, so use judgment about callable routines to know when to similarly include a sub playbook.
+Some examples of good times to do so are
+* When a set of multiple commands share a single `when` conditional
+* When a set of multiple commands are being looped together over a list of items
+* When a single large role is doing many complicated tasks and cannot easily be broken into multiple roles, but the process proceeds in multiple related stages
+====
++
+* Avoid calling the `package` module iteratively with the `{{ item }}` argument, as this is impressively more slow than calling it with the line `name: "{{ foo_packages }}"`.
+  The same can go for many other modules that can be given an entire list of items all at once.
+* Use meta modules when possible.
++
+[%collapsible]
+====
+Rationale:: This will allow our playbooks to run on the widest selection of operating systems possible without having to modify any more tasks than is necessary.
+Examples::
+* Instead of using the `upstart` and `systemd` modules, use the `service`
+module when at all possible.
+* Similarly for package management, use `package` instead of `yum` or `dnf` or
+similar.
+====
++
+* Avoid the use of `lineinfile` wherever that might be feasible.
++
+[%collapsible]
+====
+Rationale:: Slight miscalculations in how it is used can lead to a loss of idempotence.
+Modifying config files with it can cause the Ansible code to become arcane and difficult to read, especially for someone not familiar with the file in question.
+Try editing files directly using other built-in modules (e.g. `ini_file`, `blockinfile`, `xml`), or reading and parsing.
+If you are modifying more than a tiny number of lines or in a manner more than trivially complex, try leveraging the `template` module, instead.
+This will allow the entire structure of the file to be seen by later users and maintainers.
+The use of `lineinfile` should include a comment with justification.
+====
++
+* Limit use of the `copy` module to copying remote files and to uploading binary blobs.
+  For all other file pushes, use the `template` module. Even if there is nothing in the file that is being templated at the current moment, having the file handled by the `template` module now makes adding that functionality much simpler than if the file is initially handled by the `copy` and then needs to be moved before it can be edited.
+# TODO: I am not convinced about the following rule:
```

Comment:
We should separate the reshuffling from changing rules, there are also a few other places which don't convince me, but let's have this discussion in separate PRs. Create an issue to start the discussion.

Reply:
Agreed. 

I removed the `TODO` line and opened https://github.com/redhat-cop/automation-good-practices/issues/16